File: December 2022/Day 27/JAVA Code/Container With Most Water.java
// Problem Link :- https://practice.geeksforgeeks.org/problems/container-with-most-water0535/1
// Video Solution Link :- 

//{ Driver Code Starts
//Initial Template for Java


import java.io.*;
import java.util.*;
import java.lang.*;


class GFG {
	public static void main (String[] args) throws IOException{
	    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int t = Integer.parseInt(br.readLine().trim()); //Inputting the testcases
		while(t-->0){
		    int n = Integer.parseInt(br.readLine()); // input size of array
		    int arr[] = new int[n];
		    String inputLine[] = br.readLine().trim().split("\\s+");
		    for(int i=0; i<n; i++){
		        arr[i] = Integer.parseInt(inputLine[i]); // input elements of array
		    }
		    Solve T = new Solve();
            long ans = T.maxArea(arr,n);
            System.out.println(ans);
		}
	}
}

// } Driver Code Ends


//User function Template for Java


class Solve{
    // Time Complexity :- O(n*100) => O(n);
    // Space Complexity :- O(1);
    long maxArea(int A[], int len){
        // Code Here
        long ans=0;
        int left,right,width;
        for(int height=1;height<=100;height++) {
            left=0;
            for(int i=0;i<len;i++) {
                if(A[i]>=height) {
                    left=i;
                    break;
                }
            }
            
            right=0;
            for(int i=len-1;i>=0;i--) {
                if(A[i]>=height) {
                    right=i;
                    break;
                }
            }
            width = right-left;
            ans = Math.max(ans,height*width);
        }
        
        return ans;
    }
    
}

File: December 2022/Day 27/C++ Code/Container With Most Water.cpp
// Problem Link :- https://practice.geeksforgeeks.org/problems/container-with-most-water0535/1
// Video Solution Link :- https://youtu.be/9IS2erZcotg

//{ Driver Code Starts
//Initial template for C++

#include<bits/stdc++.h>
using namespace std;
// int maxArea(int A[], int len);

// } Driver Code Ends
//User function template for C++

// Time Complexity :- O(100*N) => O(n);
// Space Complexity :- O(1);
long long maxArea(long long A[], int len)
{
    // Your code goes here
    long long ans=0;
    long long left,right,width;
    for(int height=1;height<=100;height++) {
        left=0;
        for(int i=0;i<len;i++) {
            if(A[i]>=height) {
                left=i;
                break;
            }
        }
        
        right=0;
        for(int i=len-1;i>=0;i--) {
            if(A[i]>=height) {
                right=i;
                break;
            }
        }
        width = right-left;
        ans = max(ans,height*width);
    }
    
    return ans;
}

//{ Driver Code Starts.

// Driver code
int main()
{
	int t;
	cin>>t;
	while(t--)
    {
        int n;
        cin>>n;
        long long arr[n];
        for(int i=0;i<n;i++)
            cin>>arr[i];
        cout<<maxArea(arr,n)<<endl;
    }
return 0;
}

// } Driver Code Ends
